File: scrudbeans-api/src/main/java/com/github/manosbatsis/scrudbeans/api/mdd/model/ModelDescriptor.java
package com.github.manosbatsis.scrudbeans.api.mdd.model;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.type.TypeVariable;
import javax.lang.model.util.Types;
import javax.persistence.EmbeddedId;
import javax.persistence.Entity;
import javax.persistence.Id;

import com.github.manosbatsis.scrudbeans.api.mdd.ScrudModelProcessorException;
import lombok.Data;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

/**
 * Base implementation for classes describing (entity) models
 */
@Slf4j
@Data
public abstract class ModelDescriptor {

	public static final String STACK_JPA = "jpa";

	private TypeElement typeElement;

	private final Boolean jpaEntity;

	private String idType;

	private final String qualifiedName;

	private final String simpleName;

	private final String packageName;

	private final String parentPackageName;

	private Map<String, String> genericParamTypes = new HashMap<>();

	public ModelDescriptor(ProcessingEnvironment processingEnv, TypeElement typeElement) throws ScrudModelProcessorException {
		this.typeElement = typeElement;
		this.jpaEntity = typeElement.getAnnotation(Entity.class) != null;
		this.simpleName = this.typeElement.getSimpleName().toString();
		this.qualifiedName = this.typeElement.getQualifiedName().toString();
		this.packageName = this.qualifiedName.substring(0, this.qualifiedName.length() - (this.simpleName.length() + 1));
		this.parentPackageName = this.packageName.substring(0, this.packageName.lastIndexOf("."));
		Types types = processingEnv.getTypeUtils();
		scanMembers(types, typeElement);
	}

	abstract void scanMember(Types types, TypeElement currentTypeElement, Element memberElement) throws ScrudModelProcessorException;

	protected void scanMembers(Types types, TypeElement currentTypeElement) throws ScrudModelProcessorException {
		for (Element e : currentTypeElement.getEnclosedElements()) {
			this.scanMember(types, currentTypeElement, e);
		}
		currentTypeElement = asTypeElement(types, currentTypeElement.getSuperclass());
		if (!currentTypeElement.getQualifiedName().contentEquals(Object.class.getCanonicalName())) {
			scanMembers(types, currentTypeElement);
		}
	}

	protected void checkIfMemberIsId(Types types, Element e) throws ScrudModelProcessorException {
		if (e.getAnnotation(Id.class) != null || e.getAnnotation(EmbeddedId.class) != null) {
			idType = getMemberType(types, e);
		}
	}

	/**
	 * Convert the given {@link TypeMirror} to a {@link TypeElement}
	 * @param typeMirror
	 * @return
	 * @throws ScrudModelProcessorException
	 */
	protected TypeElement asTypeElement(Types types, @NonNull TypeMirror typeMirror) throws ScrudModelProcessorException {
		if (typeMirror.getKind() != TypeKind.DECLARED) {
			throw new ScrudModelProcessorException();
		}
		Element element = ((DeclaredType) typeMirror).asElement();
		if (!(element.getKind().isClass() || element.getKind().isInterface())) {
			throw new ScrudModelProcessorException();
		}
		return (TypeElement) element;
	}

	/**
	 * Inspect member and get type if field or  getter, null otherwise.
	 * @param types
	 * @param scrudModelMember
	 * @return
	 * @throws ScrudModelProcessorException
	 */
	protected String getMemberType(Types types, Element scrudModelMember) throws ScrudModelProcessorException {
		String memberType = null;
		TypeMirror typeMirror = null;
		// If member is a field
		if (scrudModelMember.getKind() == ElementKind.FIELD) {
			VariableElement ve = (VariableElement) scrudModelMember;
			typeMirror = ve.asType();
			// replace generic type variables with the current concrete type
			if (typeMirror instanceof TypeVariable) {
				typeMirror = types.asMemberOf((DeclaredType) this.typeElement.asType(), ve);
			}
		}
		// If member is a getter
		else if (scrudModelMember.getKind() == ElementKind.METHOD && scrudModelMember.getSimpleName().toString().startsWith("get")) {
			ExecutableElement ee = (ExecutableElement) scrudModelMember;
			typeMirror = ee.getReturnType();
			// replace generic type variables with the current concrete type
			if (typeMirror instanceof TypeVariable) {
				typeMirror = types.asMemberOf((DeclaredType) this.typeElement.asType(), ee);
			}

		}

		// Get class/type name
		if (typeMirror != null) {
			memberType = asTypeElement(types, typeMirror).toString();
		}

		return memberType;
	}

	public String getStack() {
		return this.jpaEntity ? STACK_JPA : "";
	}

}

File: scrudbeans-annotation-processor-kotlin/src/main/kotlin/com/github/manosbatsis/scrudbeans/processor/kotlin/descriptor/ModelDescriptor.kt
package com.github.manosbatsis.scrudbeans.processor.kotlin.descriptor

import com.github.manosbatsis.kotlin.utils.ProcessingEnvironmentAware
import com.github.manosbatsis.scrudbeans.api.mdd.ScrudModelProcessorException
import javax.persistence.Entity
import javax.annotation.processing.ProcessingEnvironment
import javax.lang.model.element.*
import javax.lang.model.element.ElementKind.FIELD
import javax.lang.model.element.ElementKind.METHOD
import javax.lang.model.type.DeclaredType
import javax.lang.model.type.TypeKind.DECLARED
import javax.lang.model.type.TypeMirror
import javax.lang.model.type.TypeVariable
import javax.lang.model.util.Types

/**
 * Base implementation for classes describing (entity) models
 */
abstract class ModelDescriptor(
    override val processingEnvironment: ProcessingEnvironment,
    val typeElement: TypeElement
) : ProcessingEnvironmentAware {

    var jpaEntity: Boolean = false
    val qualifiedName: String
    val simpleName: String
    val packageName: String
    val parentPackageName: String

    init {
        jpaEntity = typeElement.getAnnotation(Entity::class.java) != null
        simpleName = this.typeElement.simpleName.toString()
        qualifiedName = this.typeElement.qualifiedName.toString()
        packageName = qualifiedName.substring(0, qualifiedName.length - (simpleName.length + 1))
        parentPackageName = packageName.substring(0, packageName.lastIndexOf("."))
        scanMembers(processingEnvironment.typeUtils, typeElement)
        finalise()
    }

    protected fun scanMembers(types: Types, currentTypeElement: Element) {
        processingEnvironment.noteMessage {
            "scanMembers, currentTypeElement: ${currentTypeElement.simpleName}, " +
                "kind: ${currentTypeElement.kind}"
        }
        when (currentTypeElement.kind) {
            ElementKind.CLASS -> {
                val typeElement = currentTypeElement as TypeElement
                scanMembers(types, typeElement, typeElement.accessibleConstructorParameterFields())
            }
            ElementKind.CONSTRUCTOR -> {
                val constructorTypeElement = currentTypeElement as ExecutableElement
                scanMembers(types, constructorTypeElement.enclosingElement as TypeElement, constructorTypeElement.parameters)
            }
            else -> throw IllegalArgumentException("Invalid element type, expected a class or constructor")
        }
    }

    protected fun scanMembers(types: Types, currentTypeElement: TypeElement, fields: List<VariableElement>) {

        processingEnvironment.noteMessage {
            "scanMembers, currentTypeElement: ${currentTypeElement.simpleName}, " +
                "fields: ${fields.joinToString(",") { it.simpleName }}"
        }
        fields.forEach {
            scanMember(types, currentTypeElement, it)
        }
        val superTypeElement = asTypeElement(types, currentTypeElement.superclass)
        if (!superTypeElement.qualifiedName.contentEquals(Any::class.java.canonicalName)) {
            scanMembers(types, superTypeElement)
        }
    }

    abstract fun scanMember(types: Types, currentTypeElement: TypeElement, memberElement: VariableElement)

    open fun finalise() {
        /*NO-OP*/
    }

    /**
     * Convert the given [TypeMirror] to a [TypeElement]
     * @param typeMirror
     * @return
     * @throws ScrudModelProcessorException
     */
    @Throws(ScrudModelProcessorException::class)
    protected fun asTypeElement(types: Types, typeMirror: TypeMirror): TypeElement {
        if (typeMirror.kind != DECLARED) {
            throw ScrudModelProcessorException(
                "Method asTypeElement Was expecting TypeKind.DECLARED but was " + typeMirror.kind
            )
        }
        val element = (typeMirror as DeclaredType).asElement()
        if (!(element.kind.isClass || element.kind.isInterface)) {
            throw ScrudModelProcessorException("Method asTypeElement Was expecting class or interface but was " + element.kind)
        }
        return element as TypeElement
    }

    /**
     * Inspect member and get type if field or  getter, null otherwise.
     * @param types
     * @param scrudModelMember
     * @return
     * @throws ScrudModelProcessorException
     */
    protected fun getMemberType(types: Types, scrudModelMember: VariableElement): String {
        var typeMirror: TypeMirror
        // If member is a field
        if (scrudModelMember.kind == FIELD || scrudModelMember.kind == ElementKind.PARAMETER) {
            val ve = scrudModelMember as VariableElement
            typeMirror = ve.asType()
            // replace generic type variables with the current concrete type
            if (typeMirror is TypeVariable) {
                typeMirror = types.asMemberOf(typeElement.asType() as DeclaredType, ve)
            }
        } else if (scrudModelMember.kind == METHOD && scrudModelMember.simpleName.toString().startsWith("get")) {
            val ee = scrudModelMember as ExecutableElement
            typeMirror = ee.returnType
            // replace generic type variables with the current concrete type
            if (typeMirror is TypeVariable) {
                typeMirror = types.asMemberOf(typeElement.asType() as DeclaredType, ee)
            }
        } else {
            throw ScrudModelProcessorException(
                "Could not process member " + scrudModelMember + ", kind: " + scrudModelMember.kind
            )
        }

        return typeMirror.asTypeElement().asKotlinClassName().toString()
    }

    val stack: String
        get() = if (jpaEntity!!) STACK_JPA else ""

    companion object {
        const val STACK_JPA = "jpa"
    }
}